gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceExtractor.java

```diff
@@ -979,39 +1118,33 @@ private void fetchResultBatchWithRetry(RecordSetList<JsonElement> rs)
   @Override
   public void closeConnection() throws Exception {
     if (this.bulkConnection != null
-        && !this.bulkConnection.getJobStatus(this.bulkJob.getId()).getState().toString().equals("Closed")) {
+        && !this.bulkConnection.getJobStatus(this.getBulkJobId()).getState().toString().equals("Closed")) {
       log.info("Closing salesforce bulk job connection");
-      this.bulkConnection.closeJob(this.bulkJob.getId());
+      this.bulkConnection.closeJob(this.getBulkJobId());
     }
   }
 
-  public static List<Command> constructGetCommand(String restQuery) {
+  private static List<Command> constructGetCommand(String restQuery) {
     return Arrays.asList(new RestApiCommand().build(Arrays.asList(restQuery), RestApiCommandType.GET));
   }
 
   /**
    * Waits for the PK batches to complete. The wait will stop after all batches are complete or on the first failed batch
    * @param batchInfoList list of batch info
-   * @param retryInterval the polling interval
+   * @param waitInterval the polling interval
    * @return the last {@link BatchInfo} processed
    * @throws InterruptedException
    * @throws AsyncApiException
    */
-  private BatchInfo waitForPkBatches(BatchInfoList batchInfoList, int retryInterval)
+  private BatchInfo waitForPkBatches(BatchInfoList batchInfoList, int waitInterval)
       throws InterruptedException, AsyncApiException {
     BatchInfo batchInfo = null;
     BatchInfo[] batchInfos = batchInfoList.getBatchInfo();
 
-    // Wait for all batches other than the first one. The first one is not processed in PK chunking mode
-    for (int i = 1; i < batchInfos.length; i++) {
-      BatchInfo bi = batchInfos[i];
-
-      // get refreshed job status
-      bi = this.bulkConnection.getBatchInfo(this.bulkJob.getId(), bi.getId());
-
-      while ((bi.getState() != BatchStateEnum.Completed)
-          && (bi.getState() != BatchStateEnum.Failed)) {
-        Thread.sleep(retryInterval * 1000);
+    for (BatchInfo bi: batchInfos) {
+      BatchStateEnum state = bi.getState();
+      while (state != BatchStateEnum.Completed && state != BatchStateEnum.Failed && state != BatchStateEnum.NotProcessed) {
+        Thread.sleep(waitInterval * 1000);
         bi = this.bulkConnection.getBatchInfo(this.bulkJob.getId(), bi.getId());
```

Comment:
You are assigning to the outside loop variable. Not sure if this is intentional. Also, I don't see state being updated, so won't this loop forever?

Reply:
Good catch! Thanks!
I should not use the state variable. My test worked, because I was using break point, the time was enough to let sfdc execute.

I did more refactoring for this part. Pushing code advance. Will do test during today. and update you. Thanks!